### ISSUE TITLE
Operations show the source of the change

### ISSUE THREAD

**Issue by chasefarmer2808** [ianstormtaylor/slate#4178]
**Problem**
When any change occurs within the editor, there is no way to determine who or what made that change.  Was it from the user?  Programmatically though an API call?  A peer user (collaboration)?  It would be helpful to know this information when business logic is needed post-change.  Auto save is an example that comes to mind.  For auto save to work, I need to listen to changes on the editor, and only save a change when I know that I made the change.  In a collaborative setting, this is impossible to do because there is no distinction between changes I have made and ones other users have made.

**Solution**
Create a base abstract Operation type and add a source attribute to it.  Have all derived Operation types extend this abstract base type.  Then, in the default implementation for `Editor.apply`, set the value of this source attribute to a default, like "api" or "user".  QuillJS does something similar:
```
quill.on('text-change', function(delta, oldDelta, source) {
  if (source == 'api') {
    console.log("An API call triggered this change.");
  } else if (source == 'user') {
    console.log("A user action triggered this change.");
  }
});
```

**Alternatives**
Allow operation types to be extendable so a source field can be appended onto it.  Perhaps through a plugin.

**Context**
The main use cases I see this enabling is auto save and change tracking.

**Comment by juliankrispel** [comment]
@chasefarmer2808  you can already do this by overriding the `apply` method of slate.js, [see related docs](https://docs.slatejs.org/concepts/08-plugins), here's an example

```tsx
const withMetaData = editor => {
  const { apply } = editor

  editor.apply = op => {
    apply({
      ...op,
      myUserId,
    })
  }

  return editor
}
```

**Comment by chasefarmer2808** [comment]
@juliankrispel I don't think that approach will work with collaboration.  When an operation comes in from a remote user, yes it will indeed have the metadata.  But it will be overridden by this when applying it to the local editor.  I guess you could have some checks to handle that.  But I'm looking for an approach that is more built-in.

**Comment by juliankrispel** [comment]
> @juliankrispel I don't think that approach will work with collaboration. When an operation comes in from a remote user, yes it will indeed have the metadata. But it will be overridden by this when applying it to the local editor. I guess you could have some checks to handle that. But I'm looking for an approach that is more built-in.

@chasefarmer2808 it's worked for me so far and it is consistent with other slate patterns. Afaik there's nothing built-in other than that.

**Comment by chasefarmer2808** [comment]
> it's worked for me so far
Can you give a full example of this?  I am using the [slate-yjs](https://github.com/BitPhinix/slate-yjs) library for real-time collaboration.  When overridding the `apply` method, yes I can append a userId field to tell other peers who made the change.  But there are two problems with this: 
1. slate-yjs strips incoming operations of their metadata.
2. Even if it didn't, I'd have to add extra logic to prevent the overrided `apply` function on the peer's side from overriding the userId field with their own userId.  

I still think building this into Slate is the proper approach, and could enable more cool features like change tracking and auto-save.

**Comment by juliankrispel** [comment]
> Can you give a full example of this?

I can't I'm afraid. All of collaborative editing work I've done so far isn't open sourced atm. If that changes I'll gladly let you know.

> slate-yjs strips incoming operations of their metadata.

Nothing we can do about since that's a different project.

> Even if it didn't, I'd have to add extra logic to prevent the overrided apply function on the peer's side from overriding the userId field with their own userId.

Hard to help in this case without looking at the code. I imagine though that you have to treat incoming operations differently anyway.

> I still think building this into Slate is the proper approach

I think a slate library that provides good primitives for collaborative editing would be great 👍

**Comment by dylans** [comment]
The example overriding apply is nice, but doesn't cover calling a transform directly iiuc, e.g. `Transforms.setNodes`?
